dbgpt/model/proxy/llms/gemini.py

```diff
@@ -0,0 +1,67 @@
+from typing import List
+
+from dbgpt.model.proxy.llms.proxy_model import ProxyModel
+from dbgpt.core.interface.message import ModelMessage, ModelMessageRoleType
+
+GEMINI_DEFAULT_MODEL = "gemini-pro"
+
+# global history for the easy to support history
+# TODO refactor the history thing in the future
+history = []
+
+
+def gemini_generate_stream(
+    model: ProxyModel, tokenizer, params, device, context_len=2048
+):
+    """Zhipu ai, see: https://open.bigmodel.cn/dev/api#overview"""
+    model_params = model.get_params()
+    print(f"Model: {model}, model_params: {model_params}")
+    global history
+
+    # TODO proxy model use unified config?
+    proxy_api_key = model_params.proxy_api_key
+    proxyllm_backend = GEMINI_DEFAULT_MODEL or model_params.proxyllm_backend
+
+    generation_config = {
+        "temperature": 0.7,
+        "top_p": 1,
+        "top_k": 1,
+        "max_output_tokens": 2048,
+    }
+
+    safety_settings = [
+        {"category": "HARM_CATEGORY_HARASSMENT", "threshold": "BLOCK_MEDIUM_AND_ABOVE"},
+        {
+            "category": "HARM_CATEGORY_HATE_SPEECH",
+            "threshold": "BLOCK_MEDIUM_AND_ABOVE",
+        },
+        {
+            "category": "HARM_CATEGORY_SEXUALLY_EXPLICIT",
+            "threshold": "BLOCK_MEDIUM_AND_ABOVE",
+        },
+        {
+            "category": "HARM_CATEGORY_DANGEROUS_CONTENT",
+            "threshold": "BLOCK_MEDIUM_AND_ABOVE",
+        },
+    ]
+
+    import google.generativeai as genai
+
+    genai.configure(api_key=proxy_api_key)
```

Comment:
It may not be convenient to directly access generativelanguage.googleapis.com in China. Can more configurations be added here? Something like this:
```python
    import google.generativeai as genai
    from google.api_core import client_options
    # default api_endpoint="generativelanguage.googleapis.com"
    client_opts = client_options.ClientOptions(api_endpoint="my-proxy-api-endpoint.com")
    genai.configure(api_key=proxy_api_key, transport='rest', client_options=client_opts)
```

Reply:
I don't think so, if it can not use in China we can tell user use proxy in their env for now, caused there's no a better endpoint like ChatGPT for now.

Reply:
It can't work now use proxy in user's env :)
More detail: https://github.com/eosphoros-ai/DB-GPT/pull/616#discussion_r1334240495

The user can set up a reverse proxy server, similar to Nginx, as an API endpoint. Therefore, I am considering if the user provides `model_params.proxy_api_base`, we could think about modifying the configuration as follows:
```python
    import google.generativeai as genai

    if model_params.proxy_api_base:
        from google.api_core import client_options

        client_opts = client_options.ClientOptions(
            api_endpoint=model_params.proxy_api_base
        )
        genai.configure(
            api_key=proxy_api_key, transport="rest", client_options=client_opts
        )
    else:
        genai.configure(api_key=proxy_api_key)
    model = genai.GenerativeModel(
        model_name=proxyllm_backend,
        generation_config=generation_config,
        safety_settings=safety_settings,
    )
```
Following my testing, the above code can work properly.

Reply:
yes I'm concerned about safety at the regulatory level, and this proxy shouldn't be supported by us in the first place

Reply:
if user want, we support would be better;.

Reply:
Haha, the problem is that it currently cannot run with only the proxy for https://github.com/eosphoros-ai/DB-GPT/pull/616#discussion_r1334240495. Here, we are just providing more options, and whether to use it or not is up to the user's own consideration.

Reply:
OK, will change use some of your code, thanks